### DISCUSSION TITLE
Custom ThemeProvider and Theme

### DISCUSSION THREAD

**Discussion by TrevorRice** [system-ui/theme-ui#2396] [General] [answered]
Hi all!

I'm curious if there is a preferred or recommended approach for a component library or design system to expose its own wrappers around Theme UI's `ThemeProvider` and `Theme` type?

For example, suppose I want to enforce a specific `colors` structure:
```ts
interface Colors {
  primary: string;
  neutral: string;
  success: string;
  ...
}

interface MyDesignSystemTheme {
  ...
  colors: Colors;
  ...
}
```

If `MyDesignSystemThemeProvider` accepts a `theme` of type `MyDesignSystemTheme`:
```ts
interface MyDesignSystemThemeProviderProps {
  children?: ReactNode;
  theme: MyDesignSystemTheme;
}

export const MyDesignSystemThemeProvider = (props: MyDesignSystemThemeProviderProps) => {
  return <ThemeProvider theme={props.theme}>{props.children}</ThemeProvider>;
};
```

You'll receive the following TypeScript error:
```
Type 'MyDesignSystemTheme' is not assignable to type 'Theme | ((outerTheme: Theme) => Theme)'.
  Type 'MyDesignSystemTheme' is not assignable to type 'Theme'.
    Types of property 'colors' are incompatible.
      Type 'Colors' is not assignable to type 'ColorModesScale'.
        Type 'Colors' is not assignable to type 'ColorMode'.
          Index signature for type 'string' is missing in type 'Colors'.ts(2322)
index.d.ts(63, 5): The expected type comes from property 'theme' which is declared here on type '
```

This can be seen in [this CodeSandbox example](https://codesandbox.io/s/gallant-visvesvaraya-nkult0?file=/src/MyThemeProvider.tsx).

How would I be able to provide my own provider and theme interface that enforces a certain structure but doesn't result in a type mismatch when passing that theme to Theme UI's `ThemeProvider`?

**Comment by hasparus** [comment, marked as answer]
> Index signature for type 'string' is missing in type 'Colors'.ts(2322)

You can solve this by changing `Colors` to a `type` from `interface`.

[TS Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQBYFMToApQgN2ABN0o4BfOAMzxDgHIYMsBaAV2HoG4AoUSWHABK6AIYBjGABpEwsZIByEEhWq0GUeTG48eMAJ5h0cAMIQANtADOcALyIecOGCihRUfQC44VmK4B2AOa8Tv7obH6i5t6+AcGOPmzi4uhWVjF+wEG85LpZMKRUEsYAsvoAIqnAgf4Ayvq+mGiYxggJ4hbW3maWUFY8uXz+BVBFKXBllVbVdQ0FIM1YuATEpMtgNm1O4qjA5kSa-gD83iISMEokIXBMLd6TVTX1jQvM6Dm66AAeAvAd-r4JhVHrMXoscHhCCQyPYABQuCAbe7A6ZPOZNN7LKFrPAbACUdgAfA4nJoYGwoP44AAecFY1ZkW5YWwIBEbAB0TPQ5EJrNxVnZOz2B3Q-nI1IA9HTIQzCTkuEA)

**Reply by TrevorRice** [reply]
Thanks for the reply, @hasparus!

That does avoid the TS error when passing the theme to the `ThemeProvider`; however, you'd still run into issues when using [functional values](https://theme-ui.com/sx-prop#functional-values), for instance. This can be seen in an [updated CodeSandbox example](https://codesandbox.io/s/gallant-visvesvaraya-nkult0?file=/src/App.tsx) where `theme` is unaware of the custom `MyTheme` type and that `colors` and `neutral` should never be `undefined` (which is understandable — its inherent type is Theme UI's `Theme`).

```
'theme.colors' is possibly 'undefined'.
```

I guess my question is a bit more general — is there a recommended way to wrap Theme UI's `Theme` interface to enforce a certain structure? As described above, I'd like for `colors` (and other properties) to be required and have a specific shape while still reaping the TS/autocomplete benefits when using `sx`, for instance. Is that feasible?

**Reply by hasparus** [reply]
This should bring you half way: https://theme-ui.com/guides/typescript#exact-theme-type

But there's currently no supported API to fully override Theme UI types.
You should be able to provide your own type for `sx` prop if you reexport Theme UI's automatic JSX runtime with your own types. That's a bit of work though.
